File: src/Parser.java
import java.awt.geom.Point2D.Double;
import java.io.File;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;

public class Parser{
  public ArrayList<Poly> polys;
  public Double xrange, yrange;

  public Parser(){
    byte[] data = readBytes("data/large.shp");
    ArrayList<String> names = readStrings("data/names.txt");
    ArrayList<String> visited = readStrings("data/visited.txt");
    ArrayList<String> info = parseVisited(visited);
    
    ByteBuffer bb = ByteBuffer.wrap(data);
    parseHeader(bb);
    polys = parseObjects(bb);
    postParsing(polys, names, visited, info);
  }

  private ArrayList<String> parseVisited(ArrayList<String> visited){
    ArrayList<String> info = new ArrayList<String>();
    ArrayList<String> tmp = new ArrayList<String>();
    String vstr = "", res = "";
    for(String str : visited)
	vstr += str;
    String[] countries = vstr.split("#");
    for(String country : countries){
	if(country.equals(countries[0])) 
	    continue; // nothing before first hash
	String[] cities = country.split("\\|");
	tmp.add(cities[0].trim());
	info.add(country);
    }
    visited.clear();
    for(String s : tmp)
	visited.add(s);
    return info;
  }

  private byte[] readBytes(String name){
    byte[] data = null;
    try{
      File f = new File(name);
      data = new byte[(int) f.length()];
      DataInputStream s = 
	new DataInputStream(new FileInputStream(f));
      s.readFully(data);
      s.close();
    }catch(Exception e){
      e.printStackTrace();
      System.exit(1);
    }
    return data;
  }

  private ArrayList<String> readStrings(String file){
    ArrayList<String> A = new ArrayList<String>();
    String line; 
    try{
      File f = new File(file);
      BufferedReader br = new BufferedReader(new FileReader(f));
      while ((line = br.readLine()) != null)
	A.add(line);
    }catch(Exception e){
      e.printStackTrace();
      System.exit(1);
    }
    return A;
  }

  void parseHeader(ByteBuffer bb){
    int fileCode = bb.getInt();
    for(int i = 0; i < 5; i++)
      bb.getInt(); // 5 unused ints
    int fileLength = bb.getInt();
    bb.order(ByteOrder.LITTLE_ENDIAN); // changes from big to small
    int version = bb.getInt();
    int shapeType = bb.getInt();
    double minX = bb.getDouble();
    double minY = bb.getDouble();
    double maxX = bb.getDouble();
    double maxY = bb.getDouble();
    double minZ = bb.getDouble();
    double maxZ = bb.getDouble();
    double minM = bb.getDouble();
    double maxM = bb.getDouble();
    xrange = new Double(minX, maxX);
    yrange = new Double(minY, maxX);
    System.out.println("File length: " + fileLength +
		       "\nVersion: " + version +
		       "\nShape type: " + shapeType +
		       "\nMBR: "+(int)minX+", "+(int)minY+", "+
		       (int)maxX+", "+(int)maxY);
  }

  ArrayList<Poly> parseObjects(ByteBuffer bb){
    ArrayList<Poly> polys = new ArrayList<Poly>();
    for(int r = 0; r < 100000; r++){
      try{
	Poly poly = new Poly();
	poly.number = bb.getInt();
	poly.length = bb.getInt();
	poly.type = bb.getInt();
	poly.minX = bb.getDouble();
	poly.minY = bb.getDouble();
	poly.maxX = bb.getDouble();
	poly.maxY = bb.getDouble();
	int parts = bb.getInt();
	int points = bb.getInt();
	for(int i = 0; i < parts; i++)
	    poly.parts.add(bb.getInt());
	for(int i = 0; i < points; i++){
	    double x = bb.getDouble();
	    double y = bb.getDouble();
	    poly.points.add(new Double(x,y));
	    poly.center.x += x;
	    poly.center.y += y;
	}
	poly.center.x /= points;
	poly.center.y /= points;
	polys.add(poly);
	// try to parse everyhing as a polyline/polygon
      }catch(Exception e){}
    }
    System.out.println(polys.size() + " objects read (max 100k)");
    return polys;
  }

  void postParsing(ArrayList<Poly> polys, 
		   ArrayList<String> names,
		   ArrayList<String> visited,
		   ArrayList<String> info){
    if(polys.size() != names.size()){
      System.out.println("Shapefile does not match names.txt.");
      System.exit(1);
    }
    // set name and visited for all polys
    for(int i = 0; i < polys.size(); i++){
	polys.get(i).name = names.get(i);
	int index = visited.indexOf(names.get(i));
	if(index != -1){
	    String cities[] = info.get(index).split("\\|");
	    for(String city : cities){
		if(city.equals(cities[0]))
		    continue;
		String str[] = city.split("\\*");
		if(str.length == 4){
		    double x = java.lang.Double.parseDouble(str[1]);
		    double y = java.lang.Double.parseDouble(str[2]);
		    polys.get(i).cities.add
			(new City(str[0], x, y, str[3]));
		}else if(str.length == 2){
		    polys.get(i).cities.add
			(new City(str[0], 0, 0, str[1]));
		}
	    }
	}
    }
    // check all records are of the same type
    boolean sameType = true;
    int whichType = polys.get(0).type;
    for(int i = 0; i < polys.size(); i++)
      if(polys.get(i).type != whichType)
	sameType = false;
    // print info
    System.out.println("Are all objects of type " + whichType +
		       "? " + (sameType?"YES":"NO"));
    System.out.println("> starting gui.");
  }
}

File: jsport/Parser.js
export class Parser {
    constructor() {
        this.polys = [];
        this.xrange = [0, 0];
        this.yrange = [0, 0];
        this.offset = 0;
        this.ready = this.parseFiles()
    }

    parseFiles() {
        return Promise.all([
            fetch('../data/large.shp').then(response => {
                if (!response.ok) throw new Error('Network response was not ok');
                return response.arrayBuffer();
            }),
            fetch('../data/names.txt').then(res => res.text()),
            fetch('../data/visited.txt').then(res => res.text())
        ])
            .then(([arrayBuffer, namesText, visitedText]) => {
                this.dataView = new DataView(arrayBuffer);
                this.parseHeader();
                this.polys = this.parseObjects();
                const names = namesText.split(/\r?\n/).filter(line => line.length > 0);
                const visitedArr = visitedText.split(/\r?\n/).filter(line => line.length > 0);
                const info = this.parseVisited(visitedArr);
                this.postParsing(names, visitedArr, info);
            })
            .catch(error => {
                console.error('Error loading file:', error);
            });
    }

    parseVisited(visited) {
        let fullCountryStrings = [];
        const countries = visited.join('').split('#').splice(1);
        visited.length = 0; // changed to just country names
        for (const country of countries) {
            visited.push(country.split('|')[0].trim());
            fullCountryStrings.push(country);
        }
        return fullCountryStrings;
    }

    parseHeader() {
        let offset = 0;
        const getIntBE = () => { const v = this.dataView.getInt32(offset, false); offset += 4; return v; };
        const getIntLE = () => { const v = this.dataView.getInt32(offset, true); offset += 4; return v; };
        const getDoubleLE = () => { const v = this.dataView.getFloat64(offset, true); offset += 8; return v; };

        const fileCode = getIntBE();
        for (let i = 0; i < 5; i++) getIntBE();
        const fileLength = getIntBE();
        const version = getIntLE();
        const shapeType = getIntLE();
        const minX = getDoubleLE();
        const minY = getDoubleLE();
        const maxX = getDoubleLE();
        const maxY = getDoubleLE();
        const minZ = getDoubleLE();
        const maxZ = getDoubleLE();
        const minM = getDoubleLE();
        const maxM = getDoubleLE();
        this.xrange = [minX, maxX];
        this.yrange = [minY, maxY];
        this.offset = offset;
        console.log(`File length: ${fileLength}
Version: ${version}
Shape type: ${shapeType}
MBR: ${parseInt(minX)}, ${parseInt(minY)}, ${parseInt(maxX)}, ${parseInt(maxY)}`);
    }

    parseObjects() {
        let offset = this.offset;
        const polys = [];
        for (let r = 0; r < 100000; r++) {
            try {
                offset = this.parsePoly(offset, polys);
            } catch (e) {
                break; // parse up to 100k lines or until error
            }
        }
        this.offset = offset;
        console.log(`${polys.length} objects read (max 100k)`);
        return polys;
    }

    parsePoly(offset, polys) {
        const poly = new Poly();
        poly.number = this.dataView.getInt32(offset, true); offset += 4;
        poly.length = this.dataView.getInt32(offset, true); offset += 4;
        poly.type = this.dataView.getInt32(offset, true); offset += 4;
        poly.minX = this.dataView.getFloat64(offset, true); offset += 8;
        poly.minY = this.dataView.getFloat64(offset, true); offset += 8;
        poly.maxX = this.dataView.getFloat64(offset, true); offset += 8;
        poly.maxY = this.dataView.getFloat64(offset, true); offset += 8;
        const numParts = this.dataView.getInt32(offset, true); offset += 4;
        const numPoints = this.dataView.getInt32(offset, true); offset += 4;
        for (let i = 0; i < numParts; i++) { // parts array
            poly.parts.push(this.dataView.getInt32(offset, true));
            offset += 4;
        }
        for (let i = 0; i < numPoints; i++) { // points array
            const x = this.dataView.getFloat64(offset, true); offset += 8;
            const y = this.dataView.getFloat64(offset, true); offset += 8;
            poly.points.push({ x, y });
        }
        polys.push(poly);
        return offset;
    }

    postParsing(names, visited, info) {
        console.log(this.polys);
        if (this.polys.length !== names.length) {
            console.log('Shapefile does not match names.txt.');
            return;
        }
        for (let i = 0; i < this.polys.length; i++) {
            this.polys[i].name = names[i];
            this.polys[i].cities = [];
            const index = visited.indexOf(names[i]);
            if (index === -1) continue;
            info[index].split('|').splice(1).forEach(c => this.postParsingCity(c, this.polys[i]));
        }
    }

    postParsingCity(city, poly) {
        const str = city.split('*');
        if (str.length === 4) {
            poly.cities.push({
                name: str[0],
                x: parseFloat(str[1]),
                y: parseFloat(str[2]),
                info: str[3]
            });
        } else if (str.length === 2) {
            poly.cities.push({
                name: str[0],
                x: 0,
                y: 0,
                info: str[1]
            });
        }
    }
}

class Poly {
    constructor() {
        this.number = 0;
        this.length = 0;
        this.type = 0;
        this.minX = 0;
        this.minY = 0;
        this.maxX = 0;
        this.maxY = 0;
        this.parts = [];
        this.points = [];
        this.center = { x: 0, y: 0 };
        this.name = '';
        this.cities = [];
    }
}
